src/main/java/org/truffleruby/core/regexp/RegexpNodes.java

```diff
@@ -693,13 +693,28 @@ public DynamicObject source(DynamicObject regexp) {
     @CoreMethod(names = "to_s")
     public abstract static class ToSNode extends CoreMethodArrayArgumentsNode {
 
-        @TruffleBoundary
-        @Specialization
+        @Specialization(guards = "hasString(regexp)")
+        public DynamicObject toSFast(DynamicObject regexp) {
+            return createString(Layouts.REGEXP.getString(regexp));
+        }
+
+        @Specialization(guards = "!hasString(regexp)")
         public DynamicObject toS(DynamicObject regexp) {
+            Rope byteList = createRope(regexp);
+            Layouts.REGEXP.setString(regexp, byteList);
+            return createString(byteList);
+        }
+
+        @TruffleBoundary
+        protected Rope createRope(DynamicObject regexp) {
             final ClassicRegexp classicRegexp = ClassicRegexp.newRegexp(getContext(), Layouts.REGEXP.getSource(regexp), Layouts.REGEXP.getRegex(regexp).getOptions());
-            return createString(classicRegexp.toByteList());
+            RopeBuilder byteList = classicRegexp.toByteList();
```

Comment:
This should be named `ropeBuilder` or just `builder`. It's unfortunate that some variable names weren't caught in a previous refactor.